File: apps/shared/Sources/FerniShared/BetterThanHuman/ActiveListening.swift
import SwiftUI
import Combine

// MARK: - Active Listening Engine
/// Provides real-time visual feedback during user speech.
/// Creates the rhythm of natural conversation - users feel heard moment-to-moment.
///
/// Signal mapping (from BETTER-THAN-HUMAN.md):
/// - Micro-Nod: 300-800ms pauses (barely perceptible 1.5px)
/// - Subtle Nod: 800-1500ms pauses (visible 2.5px)
/// - Visible Nod: 1500ms+ pauses (full 4px)
/// - Listening Lean: Emphasis points (-3px forward)

public class ActiveListeningEngine: ObservableObject {

    // MARK: - Published State

    @Published public private(set) var isListening: Bool = false
    @Published public private(set) var currentGesture: ListeningGesture = .none

    // MARK: - Configuration

    private enum Thresholds {
        static let microNodMin: TimeInterval = 0.3      // 300ms
        static let microNodMax: TimeInterval = 0.8      // 800ms
        static let subtleNodMin: TimeInterval = 0.8     // 800ms
        static let subtleNodMax: TimeInterval = 1.5     // 1500ms
        static let visibleNodMin: TimeInterval = 1.5    // 1500ms+
        static let contemplativeMin: TimeInterval = 2.5 // 2500ms+
    }

    private enum Timing {
        static let microNodDuration: TimeInterval = 0.18
        static let subtleNodDuration: TimeInterval = 0.22
        static let visibleNodDuration: TimeInterval = 0.28
        static let leanDuration: TimeInterval = 0.4
        static let gestureCooldown: TimeInterval = 0.5  // Prevent rapid-fire gestures
    }

    // MARK: - State

    private var lastGestureTime: Date = .distantPast
    private var currentPauseDuration: TimeInterval = 0
    private var pauseTimer: Timer?

    // MARK: - Public API

    public init() {}

    /// Start active listening mode
    public func startListening() {
        isListening = true
        currentPauseDuration = 0
    }

    /// Stop active listening mode
    public func stopListening() {
        isListening = false
        currentGesture = .none
        currentPauseDuration = 0
        pauseTimer?.invalidate()
        pauseTimer = nil
    }

    /// Update with current pause duration (called continuously during speech)
    public func updatePauseDuration(_ duration: TimeInterval) {
        guard isListening else { return }

        let previousDuration = currentPauseDuration
        currentPauseDuration = duration

        // Check for gesture triggers at threshold crossings
        checkGestureTriggers(previous: previousDuration, current: duration)
    }

    // MARK: - Gesture Logic

    private func checkGestureTriggers(previous: TimeInterval, current: TimeInterval) {
        // Prevent rapid-fire gestures
        let timeSinceLastGesture = Date().timeIntervalSince(lastGestureTime)
        guard timeSinceLastGesture > Timing.gestureCooldown else { return }

        // Trigger on threshold crossings (not continuously)
        if previous < Thresholds.microNodMin && current >= Thresholds.microNodMin && current < Thresholds.subtleNodMin {
            triggerGesture(.microNod)
        } else if previous < Thresholds.subtleNodMin && current >= Thresholds.subtleNodMin && current < Thresholds.visibleNodMin {
            triggerGesture(.subtleNod)
        } else if previous < Thresholds.visibleNodMin && current >= Thresholds.visibleNodMin && current < Thresholds.contemplativeMin {
            triggerGesture(.visibleNod)
        } else if previous < Thresholds.contemplativeMin && current >= Thresholds.contemplativeMin {
            triggerGesture(.contemplative)
        }
    }

    private func triggerGesture(_ gesture: ListeningGesture) {
        lastGestureTime = Date()
        currentGesture = gesture

        // Auto-reset after gesture duration
        let duration = gestureDuration(for: gesture)
        DispatchQueue.main.asyncAfter(deadline: .now() + duration) { [weak self] in
            self?.currentGesture = .none
        }
    }

    private func gestureDuration(for gesture: ListeningGesture) -> TimeInterval {
        switch gesture {
        case .none: return 0
        case .microNod: return Timing.microNodDuration
        case .subtleNod: return Timing.subtleNodDuration
        case .visibleNod: return Timing.visibleNodDuration
        case .listeningLean: return Timing.leanDuration
        case .contemplative: return Timing.leanDuration
        }
    }

    /// Trigger a listening lean (for emphasis points detected from tone)
    public func triggerLean() {
        guard isListening else { return }
        triggerGesture(.listeningLean)
    }
}

// MARK: - Listening Gestures

public enum ListeningGesture: String, Equatable {
    case none
    case microNod       // Barely perceptible (1.5px)
    case subtleNod      // Visible (2.5px)
    case visibleNod     // Full nod (4px)
    case listeningLean  // Forward lean (-3px y)
    case contemplative  // Thoughtful expression shift

    /// Transform values for this gesture
    public var transform: ListeningTransform {
        switch self {
        case .none:
            return ListeningTransform()
        case .microNod:
            return ListeningTransform(translateY: 1.5, rotate: 0.3, scale: 0.998)
        case .subtleNod:
            return ListeningTransform(translateY: 2.5, rotate: 0.5, scale: 0.996)
        case .visibleNod:
            return ListeningTransform(translateY: 4.0, rotate: 0.8, scale: 0.994)
        case .listeningLean:
            return ListeningTransform(translateY: -3.0, rotate: 0, scale: 1.01)
        case .contemplative:
            return ListeningTransform(translateY: -1.5, rotate: -0.3, scale: 1.005)
        }
    }

    /// Animation duration for this gesture
    public var duration: TimeInterval {
        switch self {
        case .none: return 0
        case .microNod: return 0.18
        case .subtleNod: return 0.22
        case .visibleNod: return 0.28
        case .listeningLean: return 0.4
        case .contemplative: return 0.4
        }
    }
}

// MARK: - Transform Values

public struct ListeningTransform: Equatable {
    public var translateY: CGFloat
    public var rotate: CGFloat      // Degrees
    public var scale: CGFloat

    public init(translateY: CGFloat = 0, rotate: CGFloat = 0, scale: CGFloat = 1.0) {
        self.translateY = translateY
        self.rotate = rotate
        self.scale = scale
    }
}

File: apps/android-native/app/src/main/java/com/ferni/voice/betterthanuman/ActiveListening.kt
package com.ferni.voice.betterthanuman

import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * Active Listening Engine
 *
 * Provides real-time visual feedback during user speech.
 * Creates the rhythm of natural conversation - users feel heard moment-to-moment.
 *
 * Signal mapping (from BETTER-THAN-HUMAN.md):
 * - Micro-Nod: 300-800ms pauses (barely perceptible 1.5px)
 * - Subtle Nod: 800-1500ms pauses (visible 2.5px)
 * - Visible Nod: 1500ms+ pauses (full 4px)
 * - Listening Lean: Emphasis points (-3px forward)
 */
class ActiveListeningEngine(
    private val scope: CoroutineScope = CoroutineScope(Dispatchers.Main)
) {

    // MARK: - State

    private val _isListening = MutableStateFlow(false)
    val isListening: StateFlow<Boolean> = _isListening.asStateFlow()

    private val _currentGesture = MutableStateFlow(ListeningGesture.NONE)
    val currentGesture: StateFlow<ListeningGesture> = _currentGesture.asStateFlow()

    // MARK: - Configuration

    private object Thresholds {
        const val MICRO_NOD_MIN_MS = 300L    // 300ms
        const val MICRO_NOD_MAX_MS = 800L    // 800ms
        const val SUBTLE_NOD_MIN_MS = 800L   // 800ms
        const val SUBTLE_NOD_MAX_MS = 1500L  // 1500ms
        const val VISIBLE_NOD_MIN_MS = 1500L // 1500ms+
        const val CONTEMPLATIVE_MIN_MS = 2500L // 2500ms+
    }

    private object Timing {
        const val MICRO_NOD_DURATION_MS = 180L
        const val SUBTLE_NOD_DURATION_MS = 220L
        const val VISIBLE_NOD_DURATION_MS = 280L
        const val LEAN_DURATION_MS = 400L
        const val GESTURE_COOLDOWN_MS = 500L  // Prevent rapid-fire gestures
    }

    // MARK: - Private State

    private var lastGestureTimeMs: Long = 0
    private var currentPauseDurationMs: Long = 0
    private var gestureResetJob: Job? = null

    // MARK: - Public API

    /**
     * Start active listening mode
     */
    fun startListening() {
        _isListening.value = true
        currentPauseDurationMs = 0
    }

    /**
     * Stop active listening mode
     */
    fun stopListening() {
        _isListening.value = false
        _currentGesture.value = ListeningGesture.NONE
        currentPauseDurationMs = 0
        gestureResetJob?.cancel()
    }

    /**
     * Update with current pause duration in milliseconds (called continuously during speech)
     */
    fun updatePauseDuration(durationMs: Long) {
        if (!_isListening.value) return

        val previousDurationMs = currentPauseDurationMs
        currentPauseDurationMs = durationMs

        // Check for gesture triggers at threshold crossings
        checkGestureTriggers(previousDurationMs, durationMs)
    }

    /**
     * Trigger a listening lean (for emphasis points detected from tone)
     */
    fun triggerLean() {
        if (!_isListening.value) return
        triggerGesture(ListeningGesture.LISTENING_LEAN)
    }

    // MARK: - Gesture Logic

    private fun checkGestureTriggers(previousMs: Long, currentMs: Long) {
        // Prevent rapid-fire gestures
        val now = System.currentTimeMillis()
        if (now - lastGestureTimeMs < Timing.GESTURE_COOLDOWN_MS) return

        // Trigger on threshold crossings (not continuously)
        when {
            previousMs < Thresholds.MICRO_NOD_MIN_MS &&
                    currentMs >= Thresholds.MICRO_NOD_MIN_MS &&
                    currentMs < Thresholds.SUBTLE_NOD_MIN_MS -> {
                triggerGesture(ListeningGesture.MICRO_NOD)
            }
            previousMs < Thresholds.SUBTLE_NOD_MIN_MS &&
                    currentMs >= Thresholds.SUBTLE_NOD_MIN_MS &&
                    currentMs < Thresholds.VISIBLE_NOD_MIN_MS -> {
                triggerGesture(ListeningGesture.SUBTLE_NOD)
            }
            previousMs < Thresholds.VISIBLE_NOD_MIN_MS &&
                    currentMs >= Thresholds.VISIBLE_NOD_MIN_MS &&
                    currentMs < Thresholds.CONTEMPLATIVE_MIN_MS -> {
                triggerGesture(ListeningGesture.VISIBLE_NOD)
            }
            previousMs < Thresholds.CONTEMPLATIVE_MIN_MS &&
                    currentMs >= Thresholds.CONTEMPLATIVE_MIN_MS -> {
                triggerGesture(ListeningGesture.CONTEMPLATIVE)
            }
        }
    }

    private fun triggerGesture(gesture: ListeningGesture) {
        lastGestureTimeMs = System.currentTimeMillis()
        _currentGesture.value = gesture

        // Auto-reset after gesture duration
        gestureResetJob?.cancel()
        gestureResetJob = scope.launch {
            delay(gesture.durationMs)
            _currentGesture.value = ListeningGesture.NONE
        }
    }
}

/**
 * Listening Gestures
 *
 * Visual gestures that show Ferni is actively engaged while user speaks.
 */
enum class ListeningGesture(
    /** Duration in milliseconds */
    val durationMs: Long,
    /** Transform values for this gesture */
    val transform: ListeningTransform
) {
    NONE(
        durationMs = 0,
        transform = ListeningTransform()
    ),

    /**
     * Barely perceptible (1.5px) - 300-800ms pauses
     */
    MICRO_NOD(
        durationMs = 180,
        transform = ListeningTransform(translateY = 1.5f, rotate = 0.3f, scale = 0.998f)
    ),

    /**
     * Visible (2.5px) - 800-1500ms pauses
     */
    SUBTLE_NOD(
        durationMs = 220,
        transform = ListeningTransform(translateY = 2.5f, rotate = 0.5f, scale = 0.996f)
    ),

    /**
     * Full nod (4px) - 1500ms+ pauses
     */
    VISIBLE_NOD(
        durationMs = 280,
        transform = ListeningTransform(translateY = 4.0f, rotate = 0.8f, scale = 0.994f)
    ),

    /**
     * Forward lean (-3px y) - emphasis points
     */
    LISTENING_LEAN(
        durationMs = 400,
        transform = ListeningTransform(translateY = -3.0f, rotate = 0f, scale = 1.01f)
    ),

    /**
     * Thoughtful expression shift - 2500ms+ pauses
     */
    CONTEMPLATIVE(
        durationMs = 400,
        transform = ListeningTransform(translateY = -1.5f, rotate = -0.3f, scale = 1.005f)
    );

    /**
     * Duration in seconds for animation compatibility
     */
    val durationSeconds: Float
        get() = durationMs / 1000f
}

/**
 * Transform values for listening gestures.
 */
data class ListeningTransform(
    /** Vertical translation in dp (positive = down, negative = forward/up) */
    val translateY: Float = 0f,

    /** Rotation in degrees */
    val rotate: Float = 0f,

    /** Scale factor */
    val scale: Float = 1.0f
)
